platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -1573,58 +1573,47 @@ private void assertUsageIsMutable() {
+        // KGP continues to set the already-set value for a given usage even though it is already set
+        boolean redundantChange = current == newValue;
 
-        DeprecationLogger.deprecateBehaviour(String.format(msgTemplate, getDisplayName(), changingUsage))
-            .withAdvice("Usage should be fixed upon creation.")
-            .willBeRemovedInGradle9()
-            .withUpgradeGuideSection(8, "configurations_allowed_usage")
-            .nagUser();
-    }
+        // KGP disables `consumable` on detached configurations even though this is not necessary
+        boolean disableUsageForDetached = isDetachedConfiguration() && !newValue;
 
-    private void maybeWarnOnRedundantUsageActivation(String usage, String method) {
-        if (!isSpecialCaseOfRedundantUsageActivation()) {
-            String msgTemplate = "The %s usage is already allowed on %s.";
-            DeprecationLogger.deprecateBehaviour(String.format(msgTemplate, usage, getDisplayName()))
-                .withAdvice(String.format("Remove the call to %s, it has no effect.", method))
-                .willBeRemovedInGradle9()
-                .withUpgradeGuideSection(8, "redundant_configuration_usage_activation")
-                .nagUser();
+        // This property exists to allow KGP to test whether they have properly resolved this deprecation.
+        // This property WILL be removed without warning.
+        if ((redundantChange || disableUsageForDetached) &&
+            !Boolean.getBoolean("org.gradle.internal.deprecation.preliminary.Configuration.redundantUsageChangeWarning.enabled")
+        ) {
+            return;
         }
-    }
 
-    /**
-     * This is a temporary method that decides if a redundant usage activation is a known/supported special case,
-     * where a deprecation warning message should not be emitted.
-     * <p>
-     * These exceptions are needed to avoid spamming deprecations warnings whenever some important 3rd party plugins like
-     * Kotlin or Android are used.
-     * <p>
-     * <ol>
-     *     <li>Redundant activation of a usage of a detached configurations should NOT warn (this done by the Kotlin plugin).</li>
-     *     <li>Configurations with a legacy role should NOT warn during redundant usage activation,
-     * since users cannot create non-legacy configurations and there is no current public API for setting roles upon creation</li>
-     *     <li>All other usage changes should warn.</li>
-     * </ol>
-     *
-     * @return {@code true} if the usage change is a known special case; {@code false} otherwise
-     */
-    private boolean isSpecialCaseOfRedundantUsageActivation() {
-        return isInLegacyRole() || isDetachedConfiguration() || isPermittedConfigurationForRedundantActivation();
+        DeprecationLogger.deprecateAction(String.format("Calling %s(%b) on %s", methodName, newValue, this))
+            .withContext("This configuration's role was set upon creation should not be changed.")
```

Comment:
```suggestion
            .withContext("This configuration's role was set upon creation and its usage should not be changed.")
```

Reply:
❔ Should this go a step farther and suggest the appropriate role-locked conf to use instead (with the desired usage)?

Reply:
OK, on 3rd thought, I think this message is confusing.

> Calling setCanBeConsumed(false) on configuration 'myConf' has been deprecated.
> This will fail with an error in Gradle 9.0.
> This configuration's role was set upon creation [and its usage] should not be changed.

There's 2 cases:
1.  This is a configuration was created using one of the new factory methods.  In this case, it's not really that the method is _deprecated_, as in a changed behavior, it's more that this was never something that worked.
2. This is a detached configuration.  In this case, the name is generated and meaningless, and we'd want some other way to ID the configuration.  "The configuration's role was set upon creation" is also not really true in this case.  And it should say "usage" instead of "role" in this case, to be consistent.

I also think this should suggest a solution and nudge towards the use of the new factory methods (even if they are still incubating).

Reply:
In case 1, an error is thrown elsewhere. This warning is not meant to cover that scenario.

In case 2, I am not sure there is a better way to identify where the configuration came from? Do you have an idea?

In terms of a suggested solution to nudge towards factory methods, I don't see how that would help here. This warning ONLY shows for configurations that already have a role set at creation. When using role-locked factory methods, there may be a bug where we accidentally emit this deprecation AND fail with the usual "you cannot change the usage" error, and I will make sure we don't do that.

Reply:
There are a few scenarios:

1. User is setting a usage on a legacy configuration. This causes no warning ever.
2. User is setting a usage on a role-locked configuration made with a factory method. This throws an error elsewhere if you change the value. Or, it prints a warning (when the flag is enabled) if you redundantly call the method.
3. User is setting a usage on a configuration that has a role set upon creation, but is unlocked. Only Gradle can create these configurations, using internal APIs. These types of configurations are created for source sets. You currently _can_ call setCanBeX methods on these configurations, but should not.

Reply:
Allowing (optional) names for detached configurations was something I proposed a while ago, but we decided to not move forward with.  It would be helpful here (assuming people started naming their detached confs).  I don't know how else we could better ID these, but the name alone isn't very helpful.  Could we get the provenance of what plugin created the conf?

I was thinking a nudge away from detached configurations and towards factory methods would be helpful.